src/main/java/io/cryostat/core/net/JFRJMXConnection.java

```diff
@@ -209,6 +219,132 @@ public synchronized String getJvmId() throws IDException, IOException {
         }
     }
 
+    private Map<String, Object> parseCompositeData(CompositeData compositeData) {
+        Map<String, Object> map = new HashMap<>();
+        for (String key : compositeData.getCompositeType().keySet()) {
+            Object value = compositeData.get(key);
+            if (value instanceof CompositeData) {
+                map.put(key, parseCompositeData((CompositeData) value));
+            } else {
+                map.put(key, value);
+            }
+        }
+        return map;
+    }
+
+    private Map<Object, Object> parseTabularData(TabularData tabularData) {
+        Set<List<?>> keySet = (Set<List<?>>) tabularData.keySet();
+        Map<Object, Object> tdMap = new HashMap<>();
+        for (List<?> keys : keySet) {
+            CompositeData compositeData = tabularData.get(keys.toArray());
+            var cd = parseCompositeData(compositeData);
+            if (keys.size() == 1 && cd.size() == 2) {
+                Object actualKey = keys.get(0);
+                Map<?, ?> valueMap = (Map<?, ?>) cd;
+                if (valueMap.containsKey("key")
+                        && valueMap.containsKey("value")
+                        && valueMap.get("key").equals(actualKey)) {
+                    tdMap.put(valueMap.get("key"), valueMap.get("value"));
+                } else {
+                    tdMap.put(keys.get(0), cd);
+                }
+            } else {
+                tdMap.put(keys, cd);
+            }
+        }
+        return tdMap;
+    }
+
+    private Object parseObject(Object obj) {
+        if (obj instanceof CompositeData) {
+            CompositeData cd = (CompositeData) obj;
+            if (cd.getCompositeType().getTypeName().equals("java.lang.management.MemoryUsage"))
+                return io.cryostat.core.net.MemoryMetrics.MemoryUsage.fromMemoryUsage(
+                        MemoryUsage.from(cd));
+            return parseCompositeData(cd);
+        } else if (obj instanceof TabularData) {
+            return parseTabularData((TabularData) obj);
+        } else {
+            return obj;
+        }
+    }
+
+    private Object defaultValue(String type) {
+        switch (type) {
+            case "boolean":
+                return null;
```

Comment:
maybe false rather than null?

Reply:
I realize I don't need this at all anymore. The attributes map is no longer a part of the objects and only get used upon construction so we don't need to put extra fields when the constructor does it anyways.